server/src/sim.rs

```diff
@@ -156,7 +164,27 @@ impl Sim {
                 })
                 .collect(),
         };
-        self.graph.clear_fresh();
+        populate_fresh_nodes(&mut self.graph);
+
+        // Load all chunks around entities corresponding to clients, which correspond to entities
+        // with a "Character" component.
+        for (_, (position, _)) in self.world.query::<(&Position, &Character)>().iter() {
+            let mut nodes =
+                nearby_nodes(&self.graph, position, self.cfg.simulation_distance as f64);
+            // Sort nodes by distance to the view to prioritize loading closer data and improve early Z
+            // performance
+            let view_pos = position.local * math::origin();
+            nodes.sort_unstable_by(|&(_, ref xf_a), &(_, ref xf_b)| {
+                math::mip(&view_pos, &(xf_a * math::origin()))
+                    .partial_cmp(&math::mip(&view_pos, &(xf_b * math::origin())))
+                    .unwrap_or(std::cmp::Ordering::Less)
+            });
```

Comment:
The server doesn't care about the order in which chunks are loaded, nor does it need to load chunks that do not overlap with players. We should only load the bare minimum without sorting, and expand in the future if necessary.

Reply:
Since the idea is for the server to only load the bare necessary chunks, the idea of loading chunks synchronously may be worth entertaining. However, we would miss out on the advantage of multiple threads.

However, I don't think only loading chunks that overlap with players is sufficient. We want sufficient room for collision checking to be able to evaluate if a player can move to a specific spot, which means that we at least need to cover where the player is and where it can move in one frame. Multiple frames would allow for wiggle room if something temporarily slows down chunk generation.

Reply:
> It may be as simple as removing the simulation_distance configuration option and setting it to the max node radius plus some buffer to account for player radius and movement speed.

That's what I had in mind. The distance should be the player radius plus the distance a player could traverse in one timestep at maximum speed, maybe plus a small margin (say, 1cm) to account for rounding error.

> Multiple frames would allow for wiggle room if something temporarily slows down chunk generation.

A larger radius means an exponential increase in load, so I'd prefer keep things minimal.

> the idea of loading chunks synchronously may be worth entertaining. However, we would miss out on the advantage of multiple threads.

For a one-step movement bound to work, we need to finish loading chunks in the same step that we deem them necessary, before running character controllers for that step. This is appealing as it guarantees consistent behavior, with the server running slow under heavy load rather than behaving strangely. We'd also want to remove the limit on in-flight chunks. Parallelism can still be employed, but we would need to block the timestep until all outstanding work is complete. At that point it might not make sense to share the same loader between client and server, since their requirements are significantly different.

Alternatively, we could stick with the loose async approach, treat unloaded chunks as solid obstacles, and allow for two steps of motion rather than one, so that under good conditions the next step has all the chunks it needs in the worst case. This might make for better use experience under bad conditions, with the server still running in real time but occasional glitches when entering new territory. On the other hand, those glitches might feel much worse than a server running slowly, and the larger margin means higher load, and therefore odds of becoming CPU-bound, to begin with. I think the approach in the previous paragraph is therefore preferable.

Reply:
I agree that it makes sense to go with the synchronous approach, which will require me to make significant changes to this PR. For now, I think it makes sense to keep it single-threaded, since the exponential growth you mentioned goes both ways, so the client generates significantly more chunks than the server, making this unlikely to be a bottleneck.

To support falling (and potentially flying), any solution we have should be able to accommodate a relatively large maximum speed. IRL terminal velocity is about 60 m/s, which is 6 meters per time step (for the current defaults), which I don't expect to be large enough to cause a bottleneck.

Reply:
SGTM. Note that bottleneck risk mainly arises in the face of large numbers of clients, but a strictly serial approach should indeed be fine for a good long time.